app/Http/Controllers/Frontend/IcsController.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace App\Http\Controllers\Frontend;
+
+use App\Http\Controllers\Backend\IcsController as BackendIcsController;
+use App\Http\Controllers\Controller;
+use App\Models\User;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+
+class IcsController extends Controller
+{
+    public function renderIcs(Request $request): ?Response {
+        $validated = $request->validate([
+                                            'user_id' => ['required', 'exists:users,id'],
+                                            'token'   => ['required', 'exists:ics_tokens,token'],
+                                            'limit'   => ['nullable', 'numeric', 'gte:1', 'lte:10000'],
+                                            'from'    => ['nullable', 'date'],
+                                            'until'   => ['nullable', 'date']
+                                        ]);
+
+        $user               = User::where('id', $validated['user_id'])->firstOrFail();
+        $validated['limit'] = $validated['limit'] ?? 1000;
+        $from               = Carbon::parse($validated['from']);
+        $until              = Carbon::parse($validated['until']);
+
+        try {
+            $calendar = BackendIcsController::generateIcsCalendar(
+                user:  $user,
+                token: $validated['token'],
+                limit: (int) $validated['limit'],
+                from:  $from,
+                until: $until
+            );
+            return response($calendar->get())
+                ->header('Content-Type', 'text/calendar')
+                ->header('charset', 'utf-8');
+        } catch (ModelNotFoundException) {
+            abort(404);
+            return null; //Because PHPStorm is whining
```

Comment:
If you want to satisfy PhpStorm, maybe use`return response('blah', 404)`. In the ICS generator we can return text in my opinion.

Reply:
But with a useful text logically.